lib/model/queue.go

```diff
@@ -91,36 +134,49 @@ func (q *jobQueue) Jobs() ([]string, []string) {
 	progress := make([]string, len(q.progress))
 	copy(progress, q.progress)
 
-	queued := make([]string, len(q.queued))
-	for i := range q.queued {
-		queued[i] = q.queued[i].name
-	}
-
-	return progress, queued
-}
+	queued := make([]string, 0, q.queued.Items())
 
-func (q *jobQueue) Shuffle() {
-	q.mut.Lock()
-	defer q.mut.Unlock()
+	atFront := make(map[string]struct{}, len(q.broughtToFront))
+	for _, f := range q.broughtToFront {
+		if _, ok := atFront[f]; !ok {
+			queued = append(queued, f)
+			atFront[f] = struct{}{}
+		}
+	}
 
-	l := len(q.queued)
-	for i := range q.queued {
-		r := rand.Intn(l)
-		q.queued[i], q.queued[r] = q.queued[r], q.queued[i]
+	var it diskoverflow.Iterator
+	switch q.order {
+	case config.OrderLargestFirst, config.OrderNewestFirst:
+		it = q.queued.NewReverseIterator()
+	default:
+		it = q.queued.NewIterator()
 	}
+	v := &queueValue{}
+	for it.Next() {
+		it.Value(v)
+		f := v.string
+		if _, ok := atFront[f]; !ok {
+			if _, ok := q.handledAtFront[f]; !ok {
+				queued = append(queued, f)
+			}
+		}
+	}
+	it.Release()
+
+	return progress, queued
 }
 
+// To be called after a puller iteration finishes
 func (q *jobQueue) Reset() {
 	q.mut.Lock()
-	defer q.mut.Unlock()
-	q.progress = nil
-	q.queued = nil
+	q.queued = diskoverflow.NewSorted(q.location)
```

Comment:
What about closing the old one yada yada yada to free up the space?

Reply:
Right, and why `Reset` and not just `Close` and then `newJobQueue` on the caller side - can't remember.